lib/fsm.ex

```diff
@@ -0,0 +1,93 @@
+defmodule Fsm do
+  @moduledoc """
+  action: :at_the_stop or :on_the_way
+  door: :opened or :closed
+  route: []
+  state: {action, route, door_state, open_door_request}
+  {on_the_way, _, closed, requested} -> {at_the_stop, _, opened, not_requested}
+  {on_the_way, _, closed, not_requested} -> self
+  {on_the_way, _, closed, not_requested} -> {on_the_way, _, closed, requested}
+  {at_the_stop, _, closed, requested} -> {at_the_stop, _, opened, not_requested}
+  {at_the_stop, _, opened, requested} -> self
+  {at_the_stop, _, closed, not_requested} -> {at_the_stop, _, opened, not_requested}
+  {at_the_stop, _, opened, not_requested} -> self {at_the_stop, _, closed, not_requested} -> {on_the_way, _, closed, not_requested}
+  """
+  use GenServer
+  require ForTests
+
+  def start_link([]) do
+    GenServer.start_link(__MODULE__, [])
+  end
+
+  @impl true
+  def init([]) do
+    {:ok, []}
+  end
+
+  @impl true
+  def handle_call({:set_route, initial_route}, _from, _) do
+    {:reply, :ok, {:at_the_stop, initial_route, :closed, :not_requested}}
+  end
+
+  @impl true
+  def handle_call(:get_route_and_stop, _from, {_, route, door, open_request}) do
+    {:reply, {:ok, route}, {:at_the_stop, [], door, open_request}}
```

Comment:
Из этого вызова не может вернуться ошибка, поэтому можно без туплы, вроде.
```suggestion
    {:reply, route, {:at_the_stop, [], door, open_request}}
```